File: 31-next-permutation/31-next-permutation.py
class Solution:
    def nextPermutation(self, nums: List[int]) -> None:
        """
        Do not return anything, modify nums in-place instead.
        
        """
        if not nums or len(nums) <= 1:
            return
        
        def swap(i,j):
            nums[i], nums[j] = nums[j],nums[i]
            
        def reverseNums(i,j):
            while i < j:
                swap(i,j)
                i += 1 
                j -= 1
        
        ind =  len(nums)-2
        
        while ind >=0 and nums[ind] >= nums[ind + 1]:
            ind -= 1
        
        
        
            
        if ind >= 0:
            ind2 = len(nums) -1
            
            while nums[ind2] <= nums[ind]:
                ind2 -= 1
            
            #swap the elelemets at ind and ind2
            swap(ind,ind2)
            
        reverseNums(ind+1,len(nums) - 1)
            
            
                
                
            
            
            
            
        
        
        
        
        
         

File: 31-next-permutation/31-next-permutation.java
class Solution {
    public static void swap(int[] nums,int i,int j){
        int temp = nums[i];
        nums[i] = nums[j];
        nums[j] = temp;
    }
    
    public static void reverse(int[] nums,int i,int j){
       while( i< j){
           swap(nums,i,j);
           i+=1;
           j--;
       }
    }
    
    
    
    public void nextPermutation(int[] nums) {
        
        if (nums.length <= 1)
        return; 
        
        
        int ind = (nums.length) - 2;
        
        while (ind >= 0 && nums[ind] >= nums[ind+1]) {
            ind--;
        }
        int ind2= nums.length -1;
        if (ind >= 0){
            
            
            while (nums[ind2] <= nums[ind] ){
                ind2--;
            }
            
            swap(nums,ind,ind2);
               
        }
        
        reverse(nums,ind+1,nums.length-1);
         
            
        
            
        
        
        
        
        
    }
}